cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/SparkPipelineRunner.java

```diff
@@ -546,7 +547,12 @@ protected SparkCollection<Object> handleJoin(Map<String, SparkCollection<Object>
       Map<String, Schema> inputSchemas = new HashMap<>();
       for (String inputStageName : pipelinePhase.getStageInputs(stageName)) {
         StageSpec inputStageSpec = pipelinePhase.getStage(inputStageName);
-        inputSchemas.put(inputStageName, inputStageSpec.getOutputSchema());
+        Port outputPort = inputStageSpec.getOutputPorts().get(stageName);
+        if (outputPort == null) {
```

Comment:
in what case the outputPort will be null ?  ( Is it the case that Albert mentioned when the actual input schema is null ?) And what's the expected behavior for null schema join other schema ?

Reply:
I don't think we ever expect this to be null, but not sure so it's safer to do this check. When the schema is null we do some hacky thing where we force the user to provide the output schema of the join, then derive the input schemas based on that.

Reply:
got it, thanks for the explanation :)